File: src/ffi/alerts/read_piece.rs
#[cxx::bridge(namespace = "ltrs")]
pub(crate) mod ffi {
    #[namespace = "libtorrent"]
    extern "C++" {
        type read_piece_alert = crate::ffi::ffi::read_piece_alert;
    }
    unsafe extern "C++" {
        include!("cpp/alerts/read_piece.h");
        include!("lt-rs/src/ffi/error.rs.h");
        type Error = crate::ffi::error::ffi::Error;

        unsafe fn read_piece_alert_get_size(alert: *mut read_piece_alert) -> i32;

        unsafe fn read_piece_alert_get_error(alert: *mut read_piece_alert) -> Error;
    }
}

File: cpp/alerts/read_piece.cpp
#include "./read_piece.h"

#include "cpp/error.h"

#include <libtorrent/gzip.hpp>
#include <libtorrent/error_code.hpp>
#include <libtorrent/i2p_stream.hpp>
#include <libtorrent/socks5_stream.hpp>
#include <libtorrent/upnp.hpp>
#include <libtorrent/natpmp.hpp>

namespace ltrs {
    int read_piece_alert_get_size(lt::read_piece_alert* a) {
        return a->size;
    }

    Error read_piece_alert_get_error(lt::read_piece_alert* a) {
        return error_code_to_error(a->error);
    }
}
